pages/api/utils/pagePropsHelpers.ts

```diff
@@ -45,3 +49,56 @@ export const loadSession: GetServerSideProps<PagePropsWithSession> = async (
     },
   };
 };
+
+// It is a common pattern in getServerSideProps to need to first extract the API
+// token from the session, redirect to the login page if the API token is
+// missing, then use the API token to do custom logic and eventually return the
+// page props. This helper eliminates some of that boilerplate. Use it like this:
+//
+// export const getServerSideProps = makeGetServerSideProps<PageProps>(async (session) => {
+//   const ssrClient = makeSsrClient(session.user.apiToken);
+//   // Use the ssrClient
+//   return {
+//     props: {
+//       pageData,
+//     },
+//   };
+// })
+export const makeGetServerSideProps = <PageProps = Record<string, unknown>>(
+  getServerSidePropsFromSession: (
+    session: Session,
+    context: GetServerSidePropsContext,
+  ) => Promise<GetServerSidePropsResult<PageProps>>,
+): GetServerSideProps<PageProps & PagePropsWithSession> => {
+  // Return a getServerSideProps method
```

Comment:
It took me a long time to understand this, is there a way we can make it eaiser to read?

Reply:
> I don't know if it's possible, but could you make this into a type defined above so it's easier to read?

I can't make it much simpler because it's dependent on the `PageProps` generic, which is only available inside of `makeGetServerSideProps`.

I could do the below though. Do you think that's slightly easier to read?

```ts
type GetServerSidePropsWithSession<PageProps> = GetServerSideProps<PageProps & PagePropsWithSession>

export const makeGetServerSideProps = <PageProps = Record<string, unknown>>(
  getServerSidePropsFromSession: (
    session: Session,
    context: GetServerSidePropsContext,
  ) => Promise<GetServerSidePropsResult<PageProps>>,
): GetServerSidePropsWithSession<PageProps> => {
```

> It took me a long time to understand this, is there a way we can make it eaiser to read?

I don't blame you, functions that return functions usually aren't easy to read. Unfortunately, I can't think of another way to implement this helper in a way that reduces boilerplate when it is used. Is the fact that this is a function returning a `getServerSideProps` function what made it hard to read and understand? Or are there other factors that make it hard to understand?

I'm definitely open to ideas of making it simpler or easier to read. I tried to put in a bunch of comments, but I can certainly add more if that would be helpful.

I'm hopeful that this is complexity that we can write once, add 100% test coverage, and then forget about. But that's certainly not an excuse to neglect looking for ways to reduce the complexity where that's possible.

Reply:
> I could do the below, though. Do you think that's slightly easier to read?

It's about the same. Could you add some comments explaining why it returns a function? I think that will help make it more understandable. (I understand, but I'm thinking about other developers reading this)